src/main/scala/org/broadinstitute/dsde/rawls/expressions/SlickExpressionParsing.scala

```diff
@@ -186,91 +192,112 @@ trait SlickExpressionParser extends JavaTokenParsers {
     // and in the case of a list there will be more than one attribute record for an entity
     val attributeForNameQuery = (for {
-      entity <- queryPipeline.get
+      (rootEntityName, entity) <- queryPipeline.get
       entityAttribute <- entityAttributeQuery if entity.id === entityAttribute.entityId
       attribute <- attributeQuery if entityAttribute.attributeId === attribute.id && attribute.name === attributeName
-    } yield (entity.name, attribute)).result
-
+    } yield (rootEntityName, entity.name, attribute)).result
 
     attributeForNameQuery.map { entityWithAttributeRecs =>
-      // unmarshalAttributes requires a structure of ((entity id, attribute rec), option[entity rec]) where
-      // the optional entity rec is used for references. Since we know we are not dealing with a reference here
-      // as this is the attribute final func, we can pass in None.
-      val attributesByEntityId = attributeQuery.unmarshalAttributes(entityWithAttributeRecs.map { case (entityId, attrRec) => ((entityId, attrRec), None) })
-      val namedAttributesOnlyByEntityId = attributesByEntityId.mapValues(_.getOrElse(attributeName, AttributeNull)).toSeq
+      val byRootEnt: Map[String, Seq[(String, String, AttributeRecord)]] = entityWithAttributeRecs groupBy { case (rootEntity, lastEntity, attribRecord) => rootEntity }
 
-      // need to sort here because some of the manipulations above don't preserve order so we can't sort in the query
-      val orderedEntityNameAndAttributes = namedAttributesOnlyByEntityId.sortWith { case ((entityName1, _), (entityName2, _)) =>
-        entityName1 < entityName2
-      }
+      byRootEnt.map { case (rootEnt, attrs) =>
+        // unmarshalAttributes requires a structure of ((entity id, attribute rec), option[entity rec]) where
+        // the optional entity rec is used for references. Since we know we are not dealing with a reference here
+        // as this is the attribute final func, we can pass in None.
+        val attributesByEntityId = attributeQuery.unmarshalAttributes(attrs.map{ case(root, attrEnt, attrRec) => ((attrEnt, attrRec), None) })
 
-      orderedEntityNameAndAttributes.map { case (_, attribute) => attribute }
+        val namedAttributesOnlyByEntityId = attributesByEntityId.mapValues(_.getOrElse(attributeName, AttributeNull)).toSeq
+        // need to sort here because some of the manipulations above don't preserve order so we can't sort in the query
+        val orderedEntityNameAndAttributes = namedAttributesOnlyByEntityId.sortWith { case ((entityName1, _), (entityName2, _)) =>
+          entityName1 < entityName2
+        }
+
+        rootEnt -> orderedEntityNameAndAttributes.map { case (_, attribute) => attribute }
+      }
     }
   }
 
+  //A saner group by than Scala's.
+  private def groupByTuples[A, B]( tupleSeq: Seq[(A,B)] ): Map[A, Seq[B]] = {
+    tupleSeq groupBy { case (a, b) => a } mapValues { _.map(_._2) }
+  }
```

Comment:
@davidangb discovered that mapValues returns a view on the backing map, it does not actually evaluate the function until later which can lead to some issues. Use plain map instead.